plugins/auth/irods_auth_plugin-native/libirods_auth_plugin-native.cpp

```diff
@@ -0,0 +1,467 @@
+
+namespace
+{
+    using json = nlohmann::json;
+
+    auto request(rcComm_t* comm, const json& msg)
+    {
+        auto str = msg.dump();
+
+        bytesBuf_t inp;
+        inp.buf = (void*)str.c_str();
+        inp.len = str.size();
+
+        bytesBuf_t* resp{};
+        auto err = procApiRequest(comm, 110000, (void*)&inp, NULL, (void**)&resp, NULL );
+
+        if (err < 0) {
+            THROW(err, "failed to perform request");
+        }
+
+        return json::parse((char*)resp->buf, (char*)resp->buf + resp->len);
+    } // request
+
+    // TODO: std::vector could be a const ref
+    auto verify(const json& p, std::vector<std::string> n)
+    {
+        for(auto&& x : n) {
+            if(!p.contains(x)) {
+                return std::make_tuple(false, x);
+            }
+        }
+
+        return std::make_tuple(true, std::string{});
+    } // verify
+
+    template<typename T>
+    auto get(const std::string& n, const json& p)
+    {
+        if(!p.contains(n)) {
+            THROW(SYS_INVALID_INPUT_PARAM, fmt::format(
+                  "authentication request missing [{}] parameter", n));
+        }
+
+        return p.at(n).get<T>();
+    } // get
+
+#ifdef RODS_SERVER
+    void throw_on_insufficient_privilege_for_proxy_user(const rsComm_t* const _comm,
+                                                        int _proxy_priv)
+    {
+        // If the proxy user is a rodsadmin in the local zone, it can do whatever it wants.
+        if (_proxy_priv >= LOCAL_PRIV_USER_AUTH) {
+            return;
+        }
+
+        // A user is allowed to authenticate itself.
+        if (std::string_view{_comm->proxyUser.userName} == _comm->clientUser.userName) {
+            return;
+        }
+
+        // A proxy user with remote priviliged user authorization can only authenticate client
+        // users from the same zone.
+        if (_proxy_priv >= REMOTE_PRIV_USER_AUTH &&
+            std::string_view{_comm->proxyUser.rodsZone} == _comm->clientUser.rodsZone) {
+            return;
+        }
+
+        // This is a proxy user attempting to authenticate a client user with insufficient
+        // permissions to do so.
+        THROW(SYS_PROXYUSER_NO_PRIV, fmt::format(
+              "proxy user: [{}] with {} no priv to auth client user: [{}].",
+              _comm->proxyUser.userName, _proxy_priv, _comm->clientUser.userName));
+    } // throw_on_insufficient_privilege_for_proxy_user
+#endif
+} // anonymous namespace
+
+namespace irods
+{
+    class native_authentication : public irods::experimental::authentication_base {
+    public:
+        native_authentication()
+        {
+            add_operation(AUTH_ESTABLISH_CONTEXT,    OPERATION(rcComm_t, native_auth_establish_context));
+            add_operation(AUTH_CLIENT_AUTH_REQUEST,  OPERATION(rcComm_t, native_auth_client_request));
+            add_operation(AUTH_CLIENT_AUTH_RESPONSE, OPERATION(rcComm_t, native_auth_client_response));
+            #ifdef RODS_SERVER
```

Comment:
consider not aligning with runtime logic.

Reply:
Will do